src/compiler/scanner.ts

```diff
@@ -2384,15 +2444,979 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
             }
-
-            while (p < end && isIdentifierPart(text.charCodeAt(p), languageVersion)) {
+            const isUnterminated = !!(tokenFlags & TokenFlags.Unterminated);
+            const endOfBody = p - (isUnterminated ? 0 : 1);
+            let regExpFlags = RegularExpressionFlags.None;
+            while (p < end) {
+                const ch = text.charCodeAt(p);
+                if (!isIdentifierPart(ch, languageVersion)) {
+                    break;
+                }
+                const flag = characterToRegularExpressionFlag(String.fromCharCode(ch));
+                if (flag === undefined) {
+                    error(Diagnostics.Unknown_regular_expression_flag, p, 1);
+                }
+                else if (regExpFlags & flag) {
+                    error(Diagnostics.Duplicate_regular_expression_flag, p, 1);
+                }
+                else if (((regExpFlags | flag) & RegularExpressionFlags.UnicodeMode) === RegularExpressionFlags.UnicodeMode) {
+                    error(Diagnostics.The_Unicode_u_flag_and_the_Unicode_Sets_v_flag_cannot_be_set_simultaneously, p, 1);
+                }
+                else {
+                    regExpFlags |= flag;
+                    const availableFrom = regularExpressionFlagToFirstAvailableLanguageVersion(flag)!;
+                    if (languageVersion < availableFrom) {
+                        error(Diagnostics.This_regular_expression_flag_is_only_available_when_targeting_0_or_later, p, 1, ScriptTarget[availableFrom]);
+                    }
+                }
                 p++;
             }
-            pos = p;
+            pos = tokenStart + 1;
+            const saveTokenPos = tokenStart;
+            const saveTokenFlags = tokenFlags;
+            scanRegularExpressionWorker(text.slice(0, endOfBody), endOfBody, regExpFlags, isUnterminated);
+            if (!isUnterminated) {
+                pos = p;
+            }
+            tokenStart = saveTokenPos;
+            tokenFlags = saveTokenFlags;
             tokenValue = text.substring(tokenStart, pos);
             token = SyntaxKind.RegularExpressionLiteral;
         }
         return token;
+
+        function scanRegularExpressionWorker(text: string, end: number, regExpFlags: RegularExpressionFlags, isUnterminated: boolean) {
+            const unicodeMode = !!(regExpFlags & RegularExpressionFlags.UnicodeMode);
+            const unicodeSetsMode = !!(regExpFlags & RegularExpressionFlags.UnicodeSets);
+            let mayContainStrings = false;
+
+            let numberOfCapturingGroups = 0;
+            const groupSpecifiers = new Set<string>();
+            const groupNameReferences: (TextRange & { name: string; })[] = [];
+            const decimalEscapes: (TextRange & { value: number; })[] = [];
+            const namedCapturingGroups: Set<string>[] = [];
+
+            // Disjunction ::= Alternative ('|' Alternative)*
+            function scanDisjunction(isInGroup: boolean) {

```

Comment:
Do we need to nest these functions? All of these nested functions inside `scanRegularExpressionWorker` are closures and thus are essentially new objects allocated every time this function is called. These, along with `scanRegularExpressionWorker` itself should be lifted out of `reScanSlashToken()` if possible as it's a lot of unnecessary overhead.

Reply:
I was just aiming to group the functions logically, otherwise we will need to prefix all functions with `scanRegExp` or move them to a new file instead. I’ll lift them out right before the PR is ready to ship for easier reviewing.